File: coderbyte/medium/missing_digit.py
# Link: https://www.coderbyte.com/editor/Missing%20Digit:Python3
# Description:
#   Have the function MissingDigit(str) take the str parameter, which will be a simple mathematical formula with three numbers, a single operator (+, -, *, or /) and an equal sign (=) and return the digit that completes the equation. In one of the numbers in the equation, there will be an "x", and your program should determine what digit is missing. For example, if str is "3x + 12 = 46" then your program should output 4. The x character can appear in any of the three numbers and all three numbers will be greater than or equal to 0 and less than or equal to 100.
# Examples
#   Input: "4 - 2 = x"
#   Output: 2
#   Input: "1x0 * 12 = 1200"
#   Output: 0
# Tags:
#   Math Fundamentals, String Manipulation

def missing_digit(strParam):
    # Divide la ecuación en el lado izquierdo y el resultado
    left, result = strParam.split(" = ")

    # Verifica si el resultado contiene "?" y lo convierte en None
    if "?" in result:
        result = None
    else:
        result = int(result)

    # Define los operadores y encuentra cuál está en el lado izquierdo
    operators = ["+", "-", "*", "/"]
    operator = None
    for op in operators:
        if op in left:
            operator = op
            operand1, operand2 = left.split(f" {op} ")
            break

    # Define una función para realizar el cálculo según el operador
    def calculate(a, b, op):
        if op == "+":
            return a + b
        elif op == "-":
            return a - b
        elif op == "*":
            return a * b
        elif op == "/":
            if b == 0:
                return None
            return a / b
        return None

    # Itera a través de los dígitos del 0 al 9 para encontrar el dígito faltante
    for i in range(10):  # Revisamos todos los dígitos posibles (0-9)
        # Reemplaza "?" con el dígito actual
        replaced_operand1 = operand1.replace("?", str(i))
        replaced_operand2 = operand2.replace("?", str(i))

        try:
            # Convertimos los operandos solo si tienen "?"
            num1 = int(replaced_operand1) if "?" in operand1 else int(operand1)
            num2 = int(replaced_operand2) if "?" in operand2 else int(operand2)

            # Si el resultado contiene "?", lo calculamos para igualar al dígito faltante
            if result is None:
                if operator == "*":
                    # En el caso de multiplicación, comprobar que el valor de num1 * num2 sea el resultado esperado
                    if calculate(num1, num2, operator) == result:
                        return (
                            num2  # Devolvemos el número faltante en la multiplicación
                        )
                else:
                    if calculate(num1, num2, operator) == i:
                        return i
            # Si el resultado es un número, verificamos si el cálculo coincide
            elif calculate(num1, num2, operator) == result:
                return i
        except ValueError:
            continue  # Ignorar valores no válidos

    return None  # Si no se encuentra solución, devuelve None


# Casos de prueba
print(missing_digit("1? + 3 = 17"))  # Salida: 4
print(missing_digit("12 - ? = 5"))  # Salida: 7
print(missing_digit("42 * ? = 84"))  # Salida: 2
print(missing_digit("6 / ? = 3"))  # Salida: 2
print(missing_digit("3? + 1 = 14"))  # Salida: None
print(missing_digit("2 - ? = 0"))  # Salida: 2
print(missing_digit("10 * ? = 90"))  # Salida: 9
print(missing_digit("10 / ? = 2"))  # Salida: 5
print(missing_digit("10 / 0 = ?"))  # Salida: None

File: coderbyte/medium/missing_digit.js
// Find the missing digit in the equation
// Have the function missingDigit(str) take the str parameter, which will be a simple mathematical formula with three numbers, and set of arithmetic operators: +, -, *, and /. The formula will be missing one digit, and your goal is to find out what that digit is. For example, if str is "3 * 2 = 1?" then your program should output 4, because 3 * 2 = 6, which is the number that when placed in the third spot of the equation, will make the equation true.
// Examples
// Input: "38 * 2 = 8?" Output: 4
// Input: "?3 + 12 = 15" Output: 0

// define missing digit function in adding, subtracting, multiplying, and dividing, in multiplying consider exceptions.


function missingDigit(str) {
    // Parse the input string
    const [left, result] = str.split(" = ");

    // Define operators and locate the operator in the left-hand side
    const operators = ["+", "-", "*", "/"];
    let operator, parts;

    for (let op of operators) {
        if (left.includes(op)) {
            operator = op;
            parts = left.split(` ${op} `);
            break;
        }
    }

    // Extract the left-hand operand, right-hand operand, and the result
    const [operand1, operand2] = parts;
    const target = parseInt(result, 10);

    // Define a function to compute based on the operator
    const calculate = (a, b, op) => {
        switch (op) {
            case "+": return a + b;
            case "-": return a - b;
            case "*": return a * b; 
            if (a % b !== 0) return null;
            else if (a % b === 0) return a * b;
            else return null;
            case "/": return a / b;
                if (b === 0) return null;
            else return Math.floor(a / b);            
        }
    };

    // Loop through 0-9 to find the missing digit
    for (let i = 0; i <= 9; i++) {
        const replacedOperand1 = operand1.replace("?", i);
        const replacedOperand2 = operand2.replace("?", i);

        const num1 = parseFloat(replacedOperand1);
        const num2 = parseFloat(replacedOperand2);

        // Calculate and check if it matches the target
        if (!isNaN(num1) && calculate(num1, num2, operator) === target) {
            return i; // Found the missing digit
        }

        if (!isNaN(num2) && calculate(num1, num2, operator) === target) {
            return i; // Found the missing digit
        }
    }

    return null; // Return null if no solution found
}

    
// keep this function call here
console.log(missingDigit("1? + 3 = 17")); // Output: 4
console.log(missingDigit("12 - ? = 5"));  // Output: 7
console.log(missingDigit("4? * 2 = 84")); // Output: 2
console.log(missingDigit("3? + 12 = 15"));   // Output: null
console.log(missingDigit("10 * ? = 90"));   // Output: 9

